File: main/users/users_answers.go
package users

import (
	"fmt"
	"math/rand"
	"sort"
	"time"
)

var subjects = []string{"MATH", "BIOLOGY", "HISTORY", "PHYSICS"}
var ratings = []int{1, 2, 3, 4, 5}
var now = time.Now()
var generatedDatesMaxDiff = 21

func timeInDays(daysDiff int) time.Duration {
	return time.Duration(daysDiff) * time.Hour * 24
}

func getDay(t time.Time) string {
	return fmt.Sprintf("%d-%d-%d", t.Year(), t.Month(), t.Day())
}

func generate(answers int, usersCount int) []UserAnswer {
	result := make([]UserAnswer, answers)
	for i := 0; i < answers; i++ {
		userID := rand.Intn(usersCount) + 1
		subject := subjects[rand.Intn(len(subjects))]
		rating := ratings[rand.Intn(len(ratings))]
		duration := -timeInDays(rand.Intn(generatedDatesMaxDiff))
		date := now.Add(duration)
		answer := UserAnswer{userID: userID, subject: subject, rating: rating, date: date}
		result[i] = answer
	}
	return result
}

func Test(answers int, users int) {
	data := generate(answers, users)

	lastTwoWeeks := filterLastDays(data, 14)
	activities := generateUsersActivity(lastTwoWeeks)
	activities = filterByActivitiesCount(activities, 100)
	printActivitiesDiagram(activities)
	fmt.Println(fmt.Sprintf("Good answerers: %.2f%%", percentageAnswersOver(activities, 3.0)))
}

func percentageAnswersOver(activities []UserActivity, over float64) float64 {
	trueFor := 0.0
	for _, act := range activities {
		if act.averageRating() > over {
			trueFor++
		}
	}
	return trueFor / float64(len(activities)) * 100
}

func printActivitiesDiagram(activities []UserActivity) {
	dateRanges := generateDates()

	for _, x := range activities {
		fmt.Printf("User: %d\n", x.userID)
		for i, _ := range dateRanges {
			fmt.Printf("%s -> %d\n", dateRanges[i], len(x.days[i].answers))
		}
	}
}

func generateDates() []string {
	var result []string
	for i := 14; i > 1; i-- {
		result = append(result, getDay(now.Add(-timeInDays(i))))
	}
	return result
}

func filterByActivitiesCount(activities []UserActivity, minimumAnswers int) []UserActivity {
	var result []UserActivity
	for _, x := range activities {
		if x.answersCount() > minimumAnswers {
			result = append(result, x)
		}
	}
	return result
}

func generateUsersActivity(answers []UserAnswer) []UserActivity {
	answersPerUser := make(map[int][]UserAnswer)
	for _, answer := range answers {
		answersPerUser[answer.userID] = append(answersPerUser[answer.userID], answer)
	}

	var activities []UserActivity
	for userID, userAnswers := range answersPerUser {
		groupedByDay := answersGroupedByDay(userAnswers)

		activities = append(activities, UserActivity{userID: userID, days: groupedByDay})
	}

	return activities
}

func answersGroupedByDay(answers []UserAnswer) []UserDay {
	var result []UserDay
	byDay := make(map[string][]UserAnswer)
	for _, answer := range answers {
		date := getDay(answer.date)
		byDay[date] = append(byDay[date], answer)
	}

	for day, answersDay := range byDay {
		result = append(result, UserDay{date: day, answers: answersDay})
	}

	sort.Slice(result, func(i, j int) bool {
		return result[i].date < (result[j].date)
	})
	return result
}

func filterLastDays(answers []UserAnswer, maximumDaysDiff int) []UserAnswer {
	result := make([]UserAnswer, 0)
	from := now.Add(-timeInDays(maximumDaysDiff))

	for _, answer := range answers {
		if answer.date.After(from) {
			result = append(result, answer)
		}
	}
	return result
}

File: src/main/kotlin/main/UsersAnswers.kt
package main

import java.time.Duration
import java.time.ZonedDateTime
import java.time.temporal.ChronoUnit
import java.util.*

val subjects = listOf("MATH", "BIOLOGY", "HISTORY", "PHYSICS")
val ratings = (1..5)

val end: ZonedDateTime = ZonedDateTime.now()
val maxTime: Duration = Duration.ofDays(21)
const val MINIMUM_ANSWERS = 100

fun generate(answers: Int, usersCount: Int): List<UserAnswer> {
    val random = Random()

    val seconds = maxTime.seconds
    if (seconds > Int.MAX_VALUE) throw RuntimeException("overvalue")

    return (1..answers).map {
        UserAnswer(
            random.nextInt(usersCount),
            Answer(
                subjects.random(),
                ratings.random(),
                end.minus((1..14).random().toLong(), ChronoUnit.DAYS)
            )
        )
    }
}

fun toUserActivity(user: Int, answers: List<Answer>): UserActivity {
    val days = answers.groupBy { it.date.toLocalDate() }
        .map { UserDay(it.key, it.value) }
        .sortedBy { it.date }

    return UserActivity(user, days)
}

fun test() {
    val generateLastTwoWeeks = (14 downTo 1).map { end.minusDays(it.toLong()) }

    val data = generate(10000, 50)

    val usersActivities = data.filter { it.answer.date.isAfter(end.minusDays(14)) }
        .groupBy { it.user }
        .mapValues { it.value.map { usersAnswer -> usersAnswer.answer } }
        .map { toUserActivity(it.key, it.value) }
        .filter { it.answersCount() > MINIMUM_ANSWERS }

    usersActivities.forEach {
        println("User: ${it.user}")
        generateLastTwoWeeks.zip(it.days).forEach { day ->
            println("${day.first.toLocalDate()} -> ${day.second.answers.size}")
        }
    }

    val goodAnswersPercentage = usersActivities.percentageMatching { it.avgRating() > 3.0 }
    println("Good answerers: ${"%.2f".format(goodAnswersPercentage)}%")
}

fun List<UserActivity>.percentageMatching(precondition: (UserActivity) -> Boolean): Double {
    return this.count(precondition) / this.count().toDouble() * 100
}